### ISSUE TITLE
OMEdit allows inconsistent connect equations

### ISSUE THREAD

**Issue by christiankral** [OpenModelica/OpenModelica#10671]
### Description

From my experience I remember that incompatible connectors can be connected in OMEdit but simulation is not possible:

![image](https://github.com/OpenModelica/OpenModelica/assets/4184218/9bd7067f-2404-429a-974e-1c2d3eb83670)

```modelica
model Connection1
  Modelica.Electrical.Analog.Basic.Ground ground annotation(
    Placement(visible = true, transformation(origin = {-60, -10}, extent = {{-10, -10}, {10, 10}}, rotation = 0)));
  Modelica.Thermal.HeatTransfer.Components.HeatCapacitor heatCapacitor annotation(
    Placement(visible = true, transformation(origin = {-32, 30}, extent = {{-10, -10}, {10, 10}}, rotation = 0)));
equation
  connect(ground.p, heatCapacitor.port) annotation(
    Line(points = {{-60, 0}, {-60, 10}, {-32, 10}, {-32, 20}}, color = {0, 0, 255}));
annotation(
    uses(Modelica(version = "4.0.0")));
end Connection1;
```

The model check shows

```
[1] 10:23:17 Translation Error
[Connection1: 7:3-8:85]: The type of variables ground.p and heatCapacitor.port
are inconsistent in connect equations.
```

So this is all good. However, I found a simple case, where a model with incompatible connections can be simulated.

### Steps to Reproduce

Consider the following model:

![image](https://github.com/OpenModelica/OpenModelica/assets/4184218/919a42c8-1df2-4f15-b171-cf2616bef205)

```modelica
model Connection2
  Modelica.Mechanics.Rotational.Components.IdealRollingWheel idealRollingWheel1 annotation(
    Placement(visible = true, transformation(origin = {28, 0}, extent = {{10, -10}, {-10, 10}}, rotation = 0)));
  Modelica.Mechanics.Rotational.Sources.ConstantSpeed constantSpeed annotation(
    Placement(visible = true, transformation(origin = {-30, 0}, extent = {{-10, -10}, {10, 10}}, rotation = 0)));
equation
  connect(constantSpeed.flange, idealRollingWheel1.flangeT) annotation(
    Line(points = {{-20, 0}, {18, 0}}));
  annotation(
    uses(Modelica(version = "4.0.0")));
end Connection2;
```

When checking this model, translation warnings are triggered:

```
[1] 10:26:58 Translation Warning
[/home/data/work/Connection4.mo: 7:3-8:40]: The following equation is INCONSISTENT due to specified unit information: constantSpeed.flange.phi = idealRollingWheel1.flangeT.s

[2] 10:26:58 Translation Warning
The units of following sub-expressions need to be equal:
- sub-expression "idealRollingWheel1.flangeT.s" has unit "m"
- sub-expression "constantSpeed.flange.phi" has unit "1"

[3] 10:26:58 Translation Warning
[Modelica.Mechanics.Rotational.Interfaces.Flange: 5:3-5:48]: [The following equation is INCONSISTENT due to specified unit information: constantSpeed.flange.tau + idealRollingWheel1.flangeT.f = 0.0](omeditmessagesbrowser:///Modelica.Mechanics.Rotational.Interfaces.Flange?lineNumber=5)

[4] 10:26:58 Translation Warning
The units of following sub-expressions need to be equal:
- sub-expression "constantSpeed.flange.tau" has unit "J"
- sub-expression "idealRollingWheel1.flangeT.f" has unit "N"

[5] 10:26:58 Scripting Notification
Check of Connection2 completed successfully.
Class Connection2 has 21 equation(s) and 21 variable(s).
18 of these are trivial equation(s).
```

Yet the model can be simulated which definitely shall not be the case. 

### Expected Behavior

An error message shall be caused the simulation shall not be allowed.

### Version and OS

```
Connected to OpenModelica 1.22.0~dev-91-gc254033
Connected to OMSimulator 2.1.1~188-gaf996ad-linux
```

### Additonal Note

When I try to modify the parameter `w_fixed` in the constant speed source an error message is caused:

![image](https://github.com/OpenModelica/OpenModelica/assets/4184218/8b8bbf17-010e-497f-beef-10bb970c35d2)

```
[1] 10:35:36 Scripting Error
Unable to set the element modifier value using command setElementModifierValue(Connection2, constantSpeed.w_fixed, $Code((1)))
```

**Comment by adeas31** [comment]
The error that you mentioned in the additional note is already fixed. Please try the latest nightly build.

Regarding the inconsistent connection equations see #4337

**Comment by casella** [comment]
> The model check shows
> 
> ```
> [1] 10:23:17 Translation Error
> [Connection1: 7:3-8:85]: The type of variables ground.p and heatCapacitor.port
> are inconsistent in connect equations.

Now that we use the new instance-based interface, I guess we could check the types already when the connection is being made, and reject it immediately, instead of doing it later at compile time.

@perost what do you think?

**Comment by casella** [comment]
@perost, I also don't get how these dimensionally incorrect connection equations are created in the first place. Shouldn't they involve variables with the same names on the connectors?

**Comment by perost** [comment]
> > The model check shows
> > ```
> > [1] 10:23:17 Translation Error
> > [Connection1: 7:3-8:85]: The type of variables ground.p and heatCapacitor.port
> > are inconsistent in connect equations.
> > ```
> 
> Now that we use the new instance-based interface, I guess we could check the types already when the connection is being made, and reject it immediately, instead of doing it later at compile time.
> 
> @perost what do you think?

Yes, see #4337.

> @perost, I also don't get how these dimensionally incorrect connection equations are created in the first place. Shouldn't they involve variables with the same names on the connectors?

I'm not sure what you mean, I don't see any dimensionally incorrect equations here. In the second example the units are not correct, but we don't consider mismatched units to be hard errors and only give warnings for that.

**Comment by casella** [comment]
@perost sorry, I messed up. Those are the types of the connectors, so the message is conceptually OK. For an end user, indicating connector names as "variables" is a bit misleading. It would be much clearer if you got something like:

> The ground.p and heatCapacitor.port connectors have different types

or something like that

**Comment by perost** [comment]
> @perost sorry, I messed up. Those are the types of the connectors, so the message is conceptually OK. For an end user, indicating connector names as "variables" is a bit misleading. It would be much clearer if you got something like:
> 
> > The ground.p and heatCapacitor.port connectors have different types
> 
> or something like that

I agree, the error message is 17 years old and could do with an update.

**Comment by casella** [comment]
Mmmh, vintage stuff :)

**Comment by perost** [comment]
#10676 improves the error message to just say that the connectors are not type compatible. The old error message is still used by the old frontend, because the old frontend adds line breaks in the message which causes issues with git due to trailing whitespace if I change it like I want.

The error could still be improved a lot though, since it doesn't actually say why the types are not compatible. But that's a lot harder to do in a user friendly way.

**Comment by casella** [comment]
> The error could still be improved a lot though, since it doesn't actually say why the types are not compatible. But that's a lot harder to do in a user friendly way.

That's quite easy to find out, obviously you can't plug a mechanical port into a thermal port 😄

**Comment by adeas31** [comment]
Fixed in #10704
